### PR TITLE
Run Hardship and Means Assessment validations via WorkflowPreProcessorService

### PR DESCRIPTION
**Pull request by karlbaker02**
This PR updates the Hardship and Means Assessment workflows to use the _WorkflowPreProcessorService_ for validating incoming requests, specifically to ensure that the application status and timestamp are valid, as well as to ensure that the user is authorised for the action that they're attempting to perform.

This validation is only triggered from the Hardship and Means Assessment workflows when the `MaatPostAssessmentProcessing` feature flag has been enabled and additionally, in the case of Means Assessment workflows, only when the Crown Court Contribution Service has been enabled.

The _FeatureDecisionService_ has been refactored as part of this PR to consolidate common logic and the tests have been parameterised to try and ensure that duplication does not become an issue as the number of methods for checking various features increase.

This PR has _not_ refactored the _WorkflowPreProcessorService_ to consolidate Rep Order retrieval into a single method, as described in the story. The reason for this is that currently, only two services (hardship orchestration and means assessment orchestration) are calling the preprocessor; between them, one of the services is using the retrieved Rep Order elsewhere in, so a call to refactor retrieval of it in the preprocessor wouldn't remove the call to retrieve it elsewhere. The Rep Order is also used to calculate the appropriate _UserActionDTO_ (in the respective calling services), which we would otherwise need to refactor for in the preprocessor. Overall, it feels like slight premature optimisation to refactor at this stage but I'm happy to hear the thoughts of others.

Note: will fix-up commits once folks are happy with the substantive changes.

[Link to story](https://dsdmoj.atlassian.net/browse/LCAM-1255)

### CODE REVIEW

File: maat-orchestration/src/main/java/uk/gov/justice/laa/crime/orchestration/service/orchestration/MeansAssessmentOrchestrationService.java [hunk 1]
```diff
@@ -28,24 +33,26 @@
...
     }
 
     public ApplicationDTO create(WorkflowRequest request) {
-
         ApplicationDTO application = request.getApplicationDTO();
         try {
             Long repId = application.getRepId();
             log.debug("Creating Means assessment for applicationId = " + repId);
+
+            preProcessRequest(request, Action.CREATE_ASSESSMENT);
```
Comment (karlbaker02) [inline maat-orchestration/src/main/java/uk/gov/justice/laa/crime/orchestration/service/orchestration/MeansAssessmentOrchestrationService.java:55]: Is this the correct action? There wasn't anything obvious in the stored procedure.

File: maat-orchestration/src/main/java/uk/gov/justice/laa/crime/orchestration/service/orchestration/MeansAssessmentOrchestrationService.java [hunk 2]
```diff
@@ -64,11 +71,12 @@ public ApplicationDTO create(WorkflowRequest request) {
     }
 
     public ApplicationDTO update(WorkflowRequest request) {
-
         ApplicationDTO application = request.getApplicationDTO();
         try {
             Long repId = application.getRepId();
             log.debug("Updating Means assessment for applicationId = " + repId);
+
+            preProcessRequest(request, Action.UPDATE_ASSESSMENT);
```
Comment (karlbaker02) [inline maat-orchestration/src/main/java/uk/gov/justice/laa/crime/orchestration/service/orchestration/MeansAssessmentOrchestrationService.java:79]: And again here, Is this the correct action? There wasn't anything obvious in the stored procedure.

File: maat-orchestration/src/main/java/uk/gov/justice/laa/crime/orchestration/service/orchestration/MeansAssessmentOrchestrationService.java (hunk 1 above)
Comment (venkat980) [reply maat-orchestration/src/main/java/uk/gov/justice/laa/crime/orchestration/service/orchestration/MeansAssessmentOrchestrationService.java:55]: this is the correct action, the stored procedures have been retired.

File: maat-orchestration/src/main/java/uk/gov/justice/laa/crime/orchestration/service/orchestration/MeansAssessmentOrchestrationService.java (hunk 2 above)
Comment (venkat980) [reply maat-orchestration/src/main/java/uk/gov/justice/laa/crime/orchestration/service/orchestration/MeansAssessmentOrchestrationService.java:79]: this is the correct action, the stored procedures have been retired. Please refer https://github.com/ministryofjustice/laa-maat-application/pull/504/files

File: maat-orchestration/src/test/java/uk/gov/justice/laa/crime/orchestration/service/FeatureDecisionServiceTest.java [hunk 3]
```diff
@@ -36,18 +46,24 @@ void givenUserDoesNotHaveFeatureToggle_whenIsC3EnabledIsInvoked_thenReturnFalse(
             .build();
 
         FeatureDecisionService featureDecisionService = new FeatureDecisionService(maatCourtDataService);
-        boolean result = featureDecisionService.isC3Enabled(request);
+        Method method = FeatureDecisionService.class.getMethod(methodName, WorkflowRequest.class);
+        boolean result = (boolean) method.invoke(featureDecisionService, request);
```
Comment (Mhowell494) [inline maat-orchestration/src/test/java/uk/gov/justice/laa/crime/orchestration/service/FeatureDecisionServiceTest.java:50]: I'm not sure I'm too keen on the use of reflection here, I feel it would probably be easier to understand and maintain if we had separate test cases for each feature.

**Review by Mhowell494 — CHANGES_REQUESTED** [review]

File: maat-orchestration/src/main/java/uk/gov/justice/laa/crime/orchestration/service/FeatureDecisionService.java [hunk 4]
```diff
@@ -25,11 +25,30 @@ public FeatureDecisionService(final MaatCourtDataService maatCourtDataService) {
...
+     * @return A boolean describing whether the MAAT post-assessment processing flow should be
+     * invoked for the user's request.
+     */
+    public boolean isMaatPostAssessmentProcessingEnabled(WorkflowRequest workflowRequest) {
+        return isFeatureEnabled(workflowRequest, CurrentFeatureToggles.MAAT_POST_ASSESSMENT_PROCESSING, FeatureToggleAction.READ);
+    }
+
+    private boolean isFeatureEnabled(
+        WorkflowRequest workflowRequest,
+        CurrentFeatureToggles featureToggle,
```
[the remark this answers is not in the document]
Comment (karlbaker02) [reply maat-orchestration/src/main/java/uk/gov/justice/laa/crime/orchestration/service/FeatureDecisionService.java:45]: Sure, happy to refactor this existing enum.

File: maat-orchestration/src/main/java/uk/gov/justice/laa/crime/orchestration/service/orchestration/HardshipOrchestrationService.java [hunk 5]
```diff
@@ -121,12 +120,23 @@ public ApplicationDTO update(WorkflowRequest request) {
...
-        validationService.isUserActionValid(userActionDTO, userSummaryDTO);
+
+        if (featureDecisionService.isMaatPostAssessmentProcessingEnabled(request)) {
+            workflowPreProcessorService.preProcessRequest(request, repOrderDTO, userActionDTO);
+        }
+        else {
+            validationService.validate(request, repOrderDTO);
+
+            UserSummaryDTO userSummaryDTO = Optional.ofNullable(userActionDTO.getUsername()).map(maatCourtDataService::getUserSummary).orElse(null);
+            validationService.isUserActionValid(userActionDTO, userSummaryDTO);
```
[the remark this answers is not in the document]
Comment (karlbaker02) [reply maat-orchestration/src/main/java/uk/gov/justice/laa/crime/orchestration/service/orchestration/HardshipOrchestrationService.java:138]: Understood and I can see where you're coming from. I'm happy to remove the feature flag and only have the validation go through the pre-processor. I was a little bit unsure here as the story mentioned that we need to replace the calls to the stored procedure with calls (via the processor) to the _ValidationService_, but in this service we are already calling the _ValidationService_ directly. I therefore wrapped in the feature flag just to ensure no other adverse or unexpected changes from moving to the pre-processor (which does slightly more work).

tl;dr: will remove the feature flag here 👍

File: maat-orchestration/src/main/java/uk/gov/justice/laa/crime/orchestration/service/orchestration/MeansAssessmentOrchestrationService.java [hunk 6]
```diff
@@ -86,30 +94,44 @@ public ApplicationDTO update(WorkflowRequest request) {
         return application;
     }
 
-    private ApplicationDTO processCrownCourtProceedings(WorkflowRequest request) {
+    private RepOrderDTO getRepOrderDTO(WorkflowRequest request) {
+        int repId = request.getApplicationDTO().getRepId().intValue();
+        return maatCourtDataService.findRepOrder(repId);
+    }
```
[the remark this answers is not in the document]
Comment (karlbaker02) [reply maat-orchestration/src/main/java/uk/gov/justice/laa/crime/orchestration/service/orchestration/MeansAssessmentOrchestrationService.java:100]: It'd be a small service but I can totally see the value, thanks for the suggestion 👍

File: maat-orchestration/src/test/java/uk/gov/justice/laa/crime/orchestration/service/WorkflowPreProcessorServiceTest.java [hunk 7]
```diff
@@ -10,80 +10,126 @@
...
+        UserActionDTO userActionDTO = TestModelDataBuilder.getUserActionDTO();
+        UserSummaryDTO userSummaryDTO = TestModelDataBuilder.getUserSummaryDTO();
+        List<String> validationErrors = List.of(
+            ValidationService.USER_DOES_NOT_HAVE_A_VALID_NEW_WORK_REASON_CODE);
+
+        when(maatCourtDataService.getUserSummary(userActionDTO.getUsername())).thenReturn(userSummaryDTO);
+        doNothing().when(validationService).validate(request, repOrderDTO);
+        doThrow(new CrimeValidationException(validationErrors)).when(validationService).isUserActionValid(userActionDTO, userSummaryDTO);
+
+        assertThrows(CrimeValidationException.class, () -> workflowPreProcessorService.preProcessRequest(request, repOrderDTO, userActionDTO));
```
[the remark this answers is not in the document]
Comment (karlbaker02) [reply maat-orchestration/src/test/java/uk/gov/justice/laa/crime/orchestration/service/WorkflowPreProcessorServiceTest.java:68]: Ah thanks, I'll double check these assertions.

File: maat-orchestration/src/test/java/uk/gov/justice/laa/crime/orchestration/service/FeatureDecisionServiceTest.java (hunk 3 above)
Comment (karlbaker02) [reply maat-orchestration/src/test/java/uk/gov/justice/laa/crime/orchestration/service/FeatureDecisionServiceTest.java:50]: This was always going to be a debatable change here :) I totally get the reasoning and I'm not a fan of magic either. The main reason I used reflection here is because the test methods for each of the individual methods are going to be the same except for the feature toggle and action, so with many methods we'd have a lot of duplication. The other benefit to the reflection approach is that we don't need to be constantly chopping and changing test methods as we add and remove the associated methods in the actual service; because the test conditions we're expecting are going to stay the same, we can simply update the methods names passed into the tests.

If we were to keep these tests as they are now, is there anything you can see which would help to make them easier to understand?

**Comment by karlbaker02** [comment]
> I agree with the plan to consolidate `RepOrder` — that’s something we can address later on.
> 
> Regarding the `HardshipOrchestrationService`, I see we’re passing `RepOrder` to the `ATS` to retrieve assessor names, but this doesn’t seem necessary since the `ATS` can already fetch that information independently. More importantly, I’m not sure we even need to track assessor names—it's unlikely this information is displayed in the UI for Crime Apply or eForms. This seems like a legacy holdover that warrants further consideration before making changes.

Thanks for the comments. Let me get back to you on what you describe about assessor names as this isn't something I've touched as part of this PR (to my knowledge).

File: maat-orchestration/src/main/java/uk/gov/justice/laa/crime/orchestration/service/FeatureDecisionService.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (karlbaker02) [reply maat-orchestration/src/main/java/uk/gov/justice/laa/crime/orchestration/service/FeatureDecisionService.java]: Done

File: maat-orchestration/src/main/java/uk/gov/justice/laa/crime/orchestration/service/orchestration/HardshipOrchestrationService.java [hunk 8]
```diff
@@ -121,12 +120,23 @@ public ApplicationDTO update(WorkflowRequest request) {
         return request.getApplicationDTO();
     }
 
-    private void validate(WorkflowRequest request, Action action, RepOrderDTO repOrderDTO) {
-        validationService.validate(request, repOrderDTO);
+    private RepOrderDTO getRepOrderDTO(WorkflowRequest request) {
+        int repId = request.getApplicationDTO().getRepId().intValue();
+        return maatCourtDataService.findRepOrder(repId);
+    }
```
[the remark this answers is not in the document]
Comment (karlbaker02) [reply maat-orchestration/src/main/java/uk/gov/justice/laa/crime/orchestration/service/orchestration/HardshipOrchestrationService.java:126]: Good spot, thanks for this! I think I probably meant to do this but obviously forgot to follow through.

[the remark this answers is not in the document]
Comment (karlbaker02) [reply maat-orchestration/src/main/java/uk/gov/justice/laa/crime/orchestration/service/orchestration/HardshipOrchestrationService.java:128]: Done

File: maat-orchestration/src/main/java/uk/gov/justice/laa/crime/orchestration/service/orchestration/MeansAssessmentOrchestrationService.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (karlbaker02) [reply maat-orchestration/src/main/java/uk/gov/justice/laa/crime/orchestration/service/orchestration/MeansAssessmentOrchestrationService.java]: Done

File: maat-orchestration/src/main/java/uk/gov/justice/laa/crime/orchestration/service/WorkflowPreProcessorService.java [hunk 9]
```diff
@@ -9,23 +10,34 @@
...
         CaseDetailDTO caseDetailsDTO = applicationDTO.getCaseDetailsDTO();
         RepStatusDTO statusDTO = applicationDTO.getStatusDTO();
+
+        validationService.validate(workflowRequest, repOrderDTO);
+
+        UserSummaryDTO userSummaryDTO = Optional.ofNullable(userActionDTO.getUsername()).map(maatCourtDataService::getUserSummary).orElse(null);
+        validationService.isUserActionValid(userActionDTO, userSummaryDTO);
+
         String caseType = caseDetailsDTO != null ? caseDetailsDTO.getCaseType() : null;
         String status = statusDTO != null ? statusDTO.getStatus() : null;
```
[the remark this answers is not in the document]
Comment (karlbaker02) [reply maat-orchestration/src/main/java/uk/gov/justice/laa/crime/orchestration/service/WorkflowPreProcessorService.java]: Done

File: maat-orchestration/src/test/java/uk/gov/justice/laa/crime/orchestration/service/WorkflowPreProcessorServiceTest.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (karlbaker02) [reply maat-orchestration/src/test/java/uk/gov/justice/laa/crime/orchestration/service/WorkflowPreProcessorServiceTest.java]: These assertions have now been switched to AssertJ 👍

File: maat-orchestration/src/test/java/uk/gov/justice/laa/crime/orchestration/service/WorkflowPreProcessorServiceTest.java [hunk 10]
```diff
@@ -10,80 +10,126 @@
...
+    private MaatCourtDataService maatCourtDataService;
+
+    @Mock
+    private ValidationService validationService;
+
     @InjectMocks
     private WorkflowPreProcessorService workflowPreProcessorService;
 
+    @Test
+    void givenAnInvalidRequest_whenWorkflowPreProcessorServiceIsInvoked_thenExceptionIsThrown() {
```
[the remark this answers is not in the document]
Comment (karlbaker02) [reply maat-orchestration/src/test/java/uk/gov/justice/laa/crime/orchestration/service/WorkflowPreProcessorServiceTest.java]: Done, hopefully that's clearer (if not then please fee free to forward suggestions :) )

File: maat-orchestration/src/test/java/uk/gov/justice/laa/crime/orchestration/service/WorkflowPreProcessorServiceTest.java [hunk 11]
```diff
@@ -10,80 +10,126 @@
...
-                        getWorkflowRequestWithCaseType(INDICTABLE.getCaseType()),
-                        TestModelDataBuilder
-                                .buildRepOrderDTO(null)),
-                Arguments.of(
-                        getWorkflowRequestWithCaseType(INDICTABLE.getCaseType()),
-                        TestModelDataBuilder
-                                .buildRepOrderDTO("RepStatus"))
+            Arguments.of(
+                TestModelDataBuilder.buildWorkFlowRequest(),
+                TestModelDataBuilder.buildRepOrderDTO(RepOrderStatus.CURR.getCode())),
```
[the remark this answers is not in the document]
Comment (karlbaker02) [reply maat-orchestration/src/test/java/uk/gov/justice/laa/crime/orchestration/service/WorkflowPreProcessorServiceTest.java:106]: Just to make sure I'm on the same page, you're saying that you would combine the test `givenApplicationStatusHasNotChanged_whenWorkflowPreProcessorServiceIsInvoked_thenWorkflowRequestIsReturned` (the one that uses these parameterised args) with the test `givenApplicationStatusHasChanged_whenWorkflowPreProcessorServiceIsInvoked_thenUpdatedWorkflowRequestWithContributionIsReturned`?

On your point about overkill, could be. Just FYI that I didn't add this in this PR, merely reformatted it to remove some of larger-than-normal indents.

File: maat-orchestration/src/main/java/uk/gov/justice/laa/crime/orchestration/mapper/MeansAssessmentMapper.java [hunk 12]
```diff
@@ -282,6 +284,25 @@ public FinancialAssessmentDTO getMeansAssessmentResponseToFinancialAssessmentDto
...
+            ? AssessmentType.FULL : AssessmentType.INIT;
+
+        return UserActionDTO.builder()
+            .username(userDTO.getUserName())
+            .sessionId(userDTO.getUserSession())
+            .newWorkReason(assessmentType == AssessmentType.INIT ?
+                NewWorkReason.getFrom(initialAssessment.getNewWorkReason().getCode()) : null)
+            .action(action)
+            .build();
+    }
```
[the remark this answers is not in the document]
Comment (karlbaker02) [reply maat-orchestration/src/main/java/uk/gov/justice/laa/crime/orchestration/mapper/MeansAssessmentMapper.java]: Thanks, done

File: maat-orchestration/src/test/java/uk/gov/justice/laa/crime/orchestration/service/WorkflowPreProcessorServiceTest.java [hunk 13]
```diff
@@ -10,80 +10,126 @@
...
 
     @ParameterizedTest
     @MethodSource("workflowRequestWithNoApplicationStatusChange")
     void givenApplicationStatusHasNotChanged_whenWorkflowPreProcessorServiceIsInvoked_thenWorkflowRequestIsReturned(WorkflowRequest request, RepOrderDTO repOrderDTO) {
-        workflowPreProcessorService.preProcessRequest(request, repOrderDTO);
+        UserActionDTO userActionDTO = TestModelDataBuilder.getUserActionDTO();
+        UserSummaryDTO userSummaryDTO = TestModelDataBuilder.getUserSummaryDTO();
+
+        when(maatCourtDataService.getUserSummary(userActionDTO.getUsername())).thenReturn(userSummaryDTO);
+        doNothing().when(validationService).validate(request, repOrderDTO);
```
[the remark this answers is not in the document]
Comment (karlbaker02) [reply maat-orchestration/src/test/java/uk/gov/justice/laa/crime/orchestration/service/WorkflowPreProcessorServiceTest.java]: Was this comment referring to the verify call of the _ValidationService_ `validate` method? If so the reason I used `doNothing` (which I've now removed anyway) is because that method returns `void` and not `boolean`.

File: maat-orchestration/src/main/java/uk/gov/justice/laa/crime/orchestration/service/orchestration/HardshipOrchestrationService.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (karlbaker02) [reply maat-orchestration/src/main/java/uk/gov/justice/laa/crime/orchestration/service/orchestration/HardshipOrchestrationService.java]: Done

**Review by Mhowell494 — APPROVED** [review]
